Fix crash when stopping the application

a1rwulf opened the pull request:
The ALSADeviceMonitor and the ALSAHControlMonitor need to be
stopped when the sink is destroyed.
Otherwise it may trigger a "device change event" when the windowing
system is destroyed and results in a segfault.
I can trigger this reliably without this fix using GBM and ALSA.

## Types of change
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves existing functionality)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

a1rwulf committed:
[AudioEngine] fix crash when stopping the application

The ALSADeviceMonitor and the ALSAHControlMonitor need to be
stopped when the sink is destroyed.
Otherwise it may trigger a "device change event" when the windowing
system is destroyed and results in a segfault.

a1rwulf commented:
@FernetMenta this is the PR I mentioned in slack

FernetMenta commented on xbmc/cores/AudioEngine/Sinks/AESinkALSA.cpp:
```diff
@@ -909,6 +909,11 @@ void CAESinkALSA::Deinitialize()
 
 void CAESinkALSA::Stop()
 {
+  m_controlMonitor.Clear();
```
Deinitialze is called quite often by AE, not only on exit

a1rwulf replied:
I could move this into the destructor.
But Clear() is already called everytime in EnumerateDevicesEx:
https://github.com/a1rwulf/xbmc/blob/8f81b6a9056639daaafe0a57c3fcc600b204136a/xbmc/cores/AudioEngine/Sinks/AESinkALSA.cpp#L1165
It will then restart the monitor some lines below.

FernetMenta replied:
I just wanted to point out that Stop i not the right place for the reason given. I don't know what clear does.

bacon-cheeseburger replied:
I was pointed towards this patch to try after filing a bug report at: https://trac.kodi.tv/ticket/17793#ticket
I haven't experienced any more segfaults when quitting Kodi since applying it so this patch seems to have fixed the bug I reported. I don't know if this will be the final version that's merged but I thought I'd share my experience.

a1rwulf replied:
Thx for your report.
Most likely it stops monitoring too often and will be revisited.

fritsch replied:
@a1rwulf what was the reason to not add it into the dtor instead?

a1rwulf replied:
Just need to do it - until know I don't know a reason why it should not work when I move it into dtor.
Will update in an hour or so.

a1rwulf commented:
@FernetMenta @fritsch I have updated the PR and moved the monitor stops into dtor.
I tried playing some files and it still seems to work fine, closing kodi works too.
Technically I wouldn't expect any difference to the prior approach as every `m_sink->Deinitialize();` in ActiveAESink.cpp is immediately followed by a `delete m_sink;`.
Still dtor is imho the better place for it.

@bacon-cheeseburger Could you test this as well and see if it's still ok?

a1rwulf commented:
jenkins build this please

bacon-cheeseburger commented:
@a1rwulf I tried your version 2 of the fix. Did 10 attempts of start kodi-wait 30 seconds-stop kodi. I did get a segfault one time with the error "ERROR: Unable to create GUI. Exiting", but I don't think that was related to this fix (I kept the crashlog). I actually had to cold boot the box to recover from that. Is there any specific testing you'd like me to do?

FernetMenta commented:
I was pinged on this and I did no ack

FernetMenta commented:
btw: this was wrong anyway because AE deletes a sink quite often.

a1rwulf commented:
Most likely hidden by the fact that Enumerate of devices starts both monitors anyway.
So what about stopping them in the Dispose-Method of ActiveAESink.cpp?

FernetMenta commented:
this is specific to ALSA. hence we don't pull it to ActiveAESink

a1rwulf commented:
Then my fix was not completely invalid imho.
I would just need to make sure that the ctor will wether start both monitor or at least make sure they are running.
This monitors are static and it's specific to the ALSA-Sink however they are started by a static method that is not connected to the ALSASink's lifetime.
ActiveAESink looks like the sink manager to me and is not allowed to control the lifetime of those monitors (lifetime = start/stop in this case).
So who would be responsible if not the manager and not the sink itself?

fritsch commented:
I acked it and I merged it as I was pinged as well. AlsaSink is responsible
for the stuff it creates.

To keep preconditions an encapsulated class has to clean up things it
creates and allocates, if it is no factory.

 The static member itself belongs to only the alsa sink. It has to clean it
latest on exit. This is basic OO fundamentals.

If the issue for argueing is management syndrome instead, just revert it.

Am 07.03.2018 10:41 vorm. schrieb "a1rwulf" <notifications@github.com>:

> Then my fix was not completely invalid imho.
> I would just need to make sure that the ctor will wether start both
> monitor or at least make sure they are running.
> This monitors are static and it's specific to the ALSA-Sink however they
> are started by a static method that is not connected to the ALSASink's
> lifetime.
> ActiveAESink looks like the sink manager to me and is not allowed to
> control the lifetime of those monitors (lifetime = start/stop in this case).
> So who would be responsible if not the manager and not the sink itself?
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/xbmc/xbmc/pull/13596#issuecomment-371081221>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABCfHbPnggaiS1gm91zO3aPCbvb82cd9ks5tb6tCgaJpZM4SVoEd>
> .
>

FernetMenta commented:
> Maybe another approach would be to add a Cleanup-Method to the IAESink interface that is empty by default and call cleanup on disposal.

I think a cleanup method fits better into the registry. Then sinks can register it with its create method.
